pkg/kubectl/cmd/label.go

```diff
@@ -98,118 +126,63 @@ func NewCmdLabel(f *cmdutil.Factory, out io.Writer) *cobra.Command {
-		if _, found := labels[removeLabel]; found {
-			return nil, nil, fmt.Errorf("can not both modify and remove a label in the same command")
-		}
-	}
-	return labels, remove, nil
-}
+// Complete adapts from the command line args and factory to the data required.
+func (o *LabelOptions) Complete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string) (err error) {
+	o.out = out
+	o.local = cmdutil.GetFlagBool(cmd, "local")
+	o.overwrite = cmdutil.GetFlagBool(cmd, "overwrite")
+	o.all = cmdutil.GetFlagBool(cmd, "all")
+	o.resourceVersion = cmdutil.GetFlagString(cmd, "resource-version")
+	o.selector = cmdutil.GetFlagString(cmd, "selector")
+	o.outputFormat = cmdutil.GetFlagString(cmd, "output")
+	o.dryrun = cmdutil.GetDryRunFlag(cmd)
 
-func labelFunc(obj runtime.Object, overwrite bool, resourceVersion string, labels map[string]string, remove []string) error {
-	accessor, err := meta.Accessor(obj)
+	resources, labelArgs, err := cmdutil.GetResourcesAndPairs(args, "label")
 	if err != nil {
 		return err
 	}
-	if !overwrite {
-		if err := validateNoOverwrites(accessor, labels); err != nil {
-			return err
-		}
-	}
-
-	objLabels := accessor.GetLabels()
-	if objLabels == nil {
-		objLabels = make(map[string]string)
-	}
+	o.resources = resources
+	o.newLabels, o.removeLabels, err = parseLabels(labelArgs)
+	return err
+}
 
-	for key, value := range labels {
-		objLabels[key] = value
+// Validate checks to the LabelOptions to see if there is sufficient information run the command.
+func (o *LabelOptions) Validate() error {
+	if len(o.resources) < 1 && cmdutil.IsFilenameEmpty(o.FilenameOptions.Filenames) {
+		return fmt.Errorf("one or more resources must be specified as <resource> <name> or <resource>/<name>")
 	}
-	for _, label := range remove {
-		delete(objLabels, label)
-	}
-	accessor.SetLabels(objLabels)
-
-	if len(resourceVersion) != 0 {
-		accessor.SetResourceVersion(resourceVersion)
+	if len(o.newLabels) < 1 && len(o.removeLabels) < 1 {
+		return fmt.Errorf("at least one label update is required")
 	}
 	return nil
 }
 
-func RunLabel(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []string, options *resource.FilenameOptions) error {
-	resources, labelArgs, err := cmdutil.GetResourcesAndPairs(args, "label")
-	if err != nil {
-		return err
-	}
-	if len(resources) < 1 && cmdutil.IsFilenameEmpty(options.Filenames) {
-		return cmdutil.UsageError(cmd, "one or more resources must be specified as <resource> <name> or <resource>/<name>")
-	}
-	if len(labelArgs) < 1 {
-		return cmdutil.UsageError(cmd, "at least one label update is required")
-	}
-
-	selector := cmdutil.GetFlagString(cmd, "selector")
-	all := cmdutil.GetFlagBool(cmd, "all")
-	overwrite := cmdutil.GetFlagBool(cmd, "overwrite")
-	resourceVersion := cmdutil.GetFlagString(cmd, "resource-version")
-	local := cmdutil.GetFlagBool(cmd, "local")
-
+// RunLabel does the work
+func (o *LabelOptions) RunLabel(f *cmdutil.Factory, cmd *cobra.Command) error {
```

Comment:
In the end state, we'd like the options struct's `Run` to stand on their own, completely divorced from cobra for re-use and delegation as well as testing.  Something to keep in mind.

Issue ref: https://github.com/kubernetes/kubernetes/issues/7311